src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -250,39 +180,71 @@ instance
+    not (ByteString bits) = ByteString $ symbolicF bits (V.parFmap (one -) ) solve
+        where
+            solve :: MonadCircuit i (BaseField c) m => Vector n i -> m (Vector n i)
+            solve xv = do
+                let xs = V.fromVector xv
+                ys <-  for xs $ \i -> newAssigned (\p -> one - p i)
+                return $ V.unsafeToVector ys
 
-    -- | Bitwise and
-    x && y = fromConstant @Natural $ toConstant x .&. toConstant y
+    ByteString l || (ByteString r) =  ByteString $ symbolic2F l r (\lv rv -> V.unsafeToVector $ Haskell.zipWith (\li ri -> li + ri - li * ri) (V.fromVector lv) (V.fromVector rv) ) solve
+        where
+            solve :: MonadCircuit i (BaseField c) m => Vector n i -> Vector n i -> m (Vector n i)
+            solve lv rv = do
+                let varsLeft = lv
+                    varsRight = rv
+                V.zipWithM  (\i j -> newAssigned $ cons i j) varsLeft varsRight
+
+            cons i j x =
+                        let xi = x i
+                            xj = x j
+                        in xi + xj - xi * xj
+
+    ByteString l && (ByteString r) = ByteString $ symbolic2F l r (\lv rv -> V.unsafeToVector $ Haskell.zipWith (*) (V.fromVector lv) (V.fromVector rv) ) solve
+        where
+            solve :: MonadCircuit i (BaseField c) m => Vector n i -> Vector n i -> m (Vector n i)
+            solve lv rv = do
+                let varsLeft = lv
+                    varsRight = rv
+                V.zipWithM  (\i j -> newAssigned (\x -> x i * x j)) varsLeft varsRight
 
-    -- | Bitwise xor
-    xor x y = fromConstant @Natural $ toConstant x `B.xor` toConstant y
+    xor (ByteString l) (ByteString r) = ByteString $ symbolic2F l r (\lv rv -> V.unsafeToVector $ Haskell.zipWith (\li ri -> li + ri - (li * ri + li * ri)) (V.fromVector lv) (V.fromVector rv) ) solve
+        where
+            solve :: MonadCircuit i (BaseField c) m => Vector n i -> Vector n i -> m (Vector n i)
+            solve lv rv = do
+                let varsLeft = lv
+                    varsRight = rv
+                V.zipWithM  (\i j -> newAssigned $ cons i j) varsLeft varsRight
 
+            cons i j x =
+                        let xi = x i
+                            xj = x j
+                        in xi + xj - (xi * xj + xi * xj)
 
 -- | A ByteString of length @n@ can only be split into words of length @wordSize@ if all of the following conditions are met:
 -- 1. @wordSize@ is not greater than @n@;
 -- 2. @wordSize@ is not zero;
 -- 3. The bytestring is not empty;
 -- 4. @wordSize@ divides @n@.
 --
+instance
+  ( KnownNat wordSize
+  , Mod n wordSize ~ 0
+  , (Div n wordSize) * wordSize ~ n
+  ) => ToWords (ByteString n (ArithmeticCircuit a)) (ByteString wordSize (ArithmeticCircuit a)) where
+
+    toWords (ByteString bits) = (\o -> ByteString $ bits { acOutput = o} ) <$> V.fromVector (V.chunks @(Div n wordSize) @wordSize $ acOutput bits)
```

Comment:
Manipulations with `acOutput` can be generalized to functions from `HApplicative` and `Package`